[5.8] Allow locks to be secure against out of order releases

janpantel opened the pull request:
Resolves: #26213 

Before this change out of order releases of the same cache lock could lead to situation where client A acquired the lock, took longer than the timeout, which made client B successfully acquire the lock. If A now finishes while B is still holding the lock A will release the lock that does not belong to it any more.

This fix introduces a unique value that is written as the cache value to stop A from deleting B's lock.

Example from the original issue:
> * Client A acquires lock `foo` with a 10 second expiration
> * Client A begins performing a task
> * Client A takes longer than expected and the lock is automatically released by the expiration
> * Client B acquires the lock
> * Client A finishes it's task and attempts to release the lock, **deleting client B's lock**
> * Client C is now able to acquire the lock

This change does not break the existing API as it introduces the `scoped($scope)` function, plus a sugar function `safe()` that generates a random `$scope` using the `uniqid()` function.

This is most likely something that should be default for 5.8 (I will create a follow up PR after this one is merged) as it makes Cache locking much more robust. For 5.7 the opt-in is important as it could otherwise disrupt already existing locks in production deployments.

Happy to discuss the API of this :-)

janpantel committed:
Add option for scoped Cache locks

Before this change out of order releases of the same cache lock could lead
to situation where client A acquired the lock, took longer than the timeout,
which made client B successfully acquire the lock. If A now finishes while B
is still holding the lock A will release the lock that does not belong to it.

This fix introduces a unique value that is written as the cache value to stop
A from deleting B's lock.

X-Coder264 commented:
This should target the master (5.8) branch as there are breaking changes (adding a method to a contract etc.).

sisve commented:
Is there any reason this isn't the new default logic? It looks like it's opt-in by calling safe().

janpantel commented:
@sisve you're right this is opt-in. This is to keep the API stable, especially true for 5.7 as the docs currently suggest using locks like:
```php
if (Cache::lock('foo', 10)->get()) {
    // Lock acquired for 10 seconds...

    Cache::lock('foo')->release();
}
```

To use the scoped logs you have to retain the lock instance. Even worse making it the default behaviour will make the example above break as the `->release()` would try to release it with a different scope value. So the example above translated would be:
```php
$lock = Cache::lock('foo', 10)->safe();
if ($lock->get()) {
    // Lock acquired for 10 seconds...

    $lock->release();
}
```

What you could also do without keeping the instance:

```php
if (Cache::lock('foo', 10)->scoped($someUniqueProcessIdentifier)->get()) {
    // Lock acquired for 10 seconds...

    Cache::lock('foo')->scoped($someUniqueProcessIdentifier)->release();
}
```

This approach could become the default in 5.8 in disregard of the drawback that it makes the API a little less pleasant. IMO it is a good trade off to make as it will make the locking way more robust.

taylorotwell commented:
Retaining a lock instance may not always be possible though? For example, you could acquire a lock during a web request, fire a queued job, when that queued job is finished you would release the lock from the job.

yuloh commented on src/Illuminate/Cache/RedisLock.php:
```diff
@@ -49,6 +49,18 @@ public function acquire()
     {
-        $this->redis->del($this->name);
+        if ($this->canRelease()) {
+            $this->redis->del($this->name);
```
There is a race condition here.  It's possible for `canRelease` to return true, the lock to then be released because it was expired, and another client to acquire the lock before `del` is called.

With Redis you can use a Lua script to perform an atomic operation.  Since scripts are blocking no other client can acquire the lock between the check and the delete.

```lua
if redis.call("get",KEYS[1]) == ARGV[1] then
    return redis.call("del",KEYS[1])
else
    return 0
end
```

Unfortunately I don't think there is a good solution for memcached.

https://redis.io/topics/distlock

janpantel replied:
Couldn't we use a transaction for that? The Redis docs say:
> It can never happen that a request issued by another client is served in the middle of the execution of a Redis transaction.

For Memcached the `cas` command can be used setting it to `null`. Though this would eliminate the possibility of storing the lock for later release as @taylorotwell pointed out above as the `cas` assumes a state on the client: http://php.net/manual/de/memcached.cas.php

yuloh replied:
> Couldn't we use a transaction for that? 

You need to check if the key matches the token before deleting it.  I don't think there is an equivalent Redis command that does that.  Redis transactions aren't like database transactions, the GET will just be queued until EXEC and you can't read the value of the key in PHP before you call `EXEC`.

janpantel committed:
Use Str::random instead of uniqid

janpantel commented on src/Illuminate/Cache/Lock.php:
```diff
@@ -99,4 +113,61 @@ public function block($seconds, $callback = null)
+     */
+    public function safe()
+    {
+        return $this->scoped(uniqid());
```
Good point, will fix

janpantel commented on src/Illuminate/Cache/Lock.php:
```diff
@@ -99,4 +113,61 @@ public function block($seconds, $callback = null)
+     */
+    protected function value()
+    {
+        return $this->isScoped() ? serialize($this->scope) : 1;
```
Actually it is, like also done in the `RedisStore`: https://github.com/laravel/framework/blob/9f313ce9bb5ad49a06ae78d33fbdd1c92a0e21f6/src/Illuminate/Cache/RedisStore.php#L276

janpantel replied:
Actually it was the test failing that was using the cache instead of the lock to verify, fixed it and got rid of the serialize.

yuloh commented:
> Retaining a lock instance may not always be possible though? For example, you could acquire a lock during a web request, fire a queued job, when that queued job is finished you would release the lock from the job.

If you needed to use the same lock instance maybe you could pass the lock token around?  i.e.

```php
// in web request
$lock = Cache::lock('foo', 10)->get();
$job = SomeJob::dispatch($lock->token());

// in worker
$lock = Cache::lock('foo', 10, $this->token)->get();
```

In this example the token parameter would be optional, and if provided would be used instead of the random token.

janpantel commented:
What @yuloh said, actually using the `->scoped($token)` function you can already do that. I think the question is whether it makes sense to bring this to 5.7 or rewrite it to target 5.8 and create a more pleasant API?

janpantel committed:
Use locks for check in test to eliminate need for serialization

janpantel committed:
Style fixes

janpantel committed:
Renamed scoped to owned

janpantel commented:
After some research I renamed the `->safe()` method to `->owned()` and changed `scope` to `owner`. This matches the usual lingo of lock packages in the OSS world.

janpantel commented:
@taylorotwell actually after trying it out I can confirm that you can dispatch a queued job (tested using the `redis` queue driver) containing the lock instance that you marked with `->owned()` before.
From the job's `handle()` function the lock can be properly released as the `owner` property is properly restored.

taylorotwell commented:
Personally I would target all of this towards 5.8 and try to write it in the best way we can.

janpantel commented:
@taylorotwell totally fine with me. I will change the PR to default to owned locks without the need of explicitly stating it and target the 5.8 branch.
Do you have any strong objections against the need of keeping the lock instance in order to be able to release it?

taylorotwell commented:
@janpantel I don't totally see a way around that if you want locks to work this way? You would at least need to know the scope. Have you researched how symfony/lock component is solving this issue?

taylorotwell commented:
It looks like `symfony/lock` uses a similar "key" approach. However, I do not see a way to release the lock from within another process like a queued job. Maybe the lock can be serialized? I'm not sure.

janpantel commented:
@taylorotwell you just wrote that when I was about to answer. Yes exactly, they also use a LUA script for Redis. So given my real world test with my 5.7 code it was possible to hand the lock into a job which gets serialised into a Redis queue. The release was successful when done from the Job instance.

My biggest struggle currently is writing an integration test that can ensure that automatically. Is there a way to start a queue listener in the integration tests?

janpantel committed:
WIP - move to 5.8 and make owned locks the default behaviour

GrahamCampbell committed:
Fixes

GrahamCampbell committed:
Fixed invalid phpdoc

GrahamCampbell committed:
Fixed invalid phpdoc

GrahamCampbell commented on src/Illuminate/Cache/Lock.php:
```diff
@@ -51,6 +65,13 @@ public function __construct($name, $seconds)
+    /**
+     * Returns the owner value written into the driver for this lock.
+     *
+     * @return mixed
```
We probably want this to be `string|null`?

janpantel replied:
@GrahamCampbell any opinion on that one?

GrahamCampbell replied:
Yes, if that is their type.

taylorotwell commented:
Eh, I don't think testing that specifically is important as simply testing the creation of a lock from an existing scope, etc.

CthulhuDen commented:
How about `Cache::block` method, will it auto-generate token as well? Also, currently (in 5.7) if closure throws the `Cache::block` will not release the lock, do you think that also deserves amending?

janpantel committed:
Restore lock with original owner token and memcached changes

taylorotwell commented:
Hmm, yeah I don't think being forced to pass the expiry time in order to release the lock makes much sense.

taylorotwell commented:
Are there plans to keep working on this?

janpantel commented:
@taylorotwell yes, just had some downtime due to the holidays. I will add a commit introducing Cache::from() later. Also I'm waiting to @GrahamCampbell to answer :-)

janpantel committed:
Introduce restoreLock function

GrahamCampbell commented:
I've replied. Please let me know if you need anything else.

janpantel committed:
Move redis lock release to lua script

janpantel committed:
Style fixes

janpantel commented:
@GrahamCampbell The only thing left is to remove the request changes if you're fine with the current state.

@taylorotwell I created the `restoreLock` function and also ported to Redis implementation to the LUA script that @yuloh mentioned. Anything else you see left tbd?

taylorotwell commented:
Can you just comment with the basic usage documentation given the current API so I can look over it and use it as a guide in my own testing?

janpantel commented:
@taylorotwell for sure. I will also go ahead and update the real docs once we have committed on the API.

To obtain a lock and release it it is now necessary to carry the instance around.

```php
if ($lock = Cache::lock('foo', 10)->get()) {
    // do stuff

    $lock->release()
}
```

If you are sure that there will be no race condition or the out of order release does not trouble you you can also use the `forceRelease` function without remembering the lock instance.

```php
if (Cache::lock('foo', 10)->get()) {
    // do stuff

    Cache::lock('foo')->forceRelease()
}
```

If you have to release the lock in another process you can get the owner token from the lock instance and restore it later.

```php
if ($lock = Cache::lock('foo', 10)->get()) {
    // do stuff

    event(new LockReleasingEvent($lock->getOwner()));
}

// later ...

$lock = Cache::restoreLock('foo', $owner);

// do stuff

$lock->release
```

A counter example, as in how to not do it would be:
```php
if (Cache::lock('foo', 10)->get()) {
    // do stuff

    Cache::lock('foo')->release()

    /*
    lock still in place as the constructor generates a random
    owner token each time a lock instance is created
    */
}
```

taylorotwell commented:
Nice work. I made the necessary changes to the new DynamoDbLock as well.

driesvints commented:
@janpantel thanks for your work on this! 👏

janpantel commented:
Thanks for making my first major contribution such a pleasant experience! :-)
